src/parser/mod.rs

```diff
@@ -9416,27 +9416,42 @@ impl<'a> Parser<'a> {
         }
     }
 
+    fn parse_set_role(&mut self, modifier: Option<Keyword>) -> Option<Statement> {
```

Comment:
thinking maybe this function should return a Result instead as usual, the current rewind behavior doesn't seem correct otherwise?

If this part throws an error, it would have already consumed some tokens from the stream.
```rust
                role_name: Some(self.parse_identifier(false).ok()?),
```
So that the assumption being made at the caller here becomes incorrect
```rust
                self.prev_token(); // replace the "role" token which was actually a variable name
```
(Can we add a test case demonstrating this expected behavior in any case?)

Thinking if `parse_set_role` is a normal parse function, then the caller can instead do `self.maybe_parse(parse_set_role)` which will do the automatic rewind if it turns out not to be a legit `SET ROLE` statement in any scenario?

Reply:
parse_identifier cannot return an error here (because of the checks above). But you are right, maybe_parse is both more robust and more readable. I changed the code.